### PR TITLE
Tensorboard path generalisation

### PR DESCRIPTION
**Pull request by bobkemp**
# Before submitting

- [ y] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [y ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [ y] Did you make sure to update the docs?   
- [ y] Did you write any new necessary tests?  

## What does this PR do?
Fixes issue 793

You can (say) specify the version on the command line or generate it from the current datetime, which I find more meaningful than "version_73".

Likewise if the tensorboard data is not going to be preserved, it's much more readable in the tensorboard gui to skip the experiment name.

## PR review
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

I learnt some things

### CODE REVIEW

File: pytorch_lightning/loggers/tensorboard.py [hunk 1]
```diff
@@ -29,9 +29,10 @@ class TensorBoardLogger(LightningLoggerBase):
 
     Args:
         save_dir (str): Save directory
-        name (str): Experiment name. Defaults to "default".
-        version (int): Experiment version. If version is not specified the logger inspects the save
+        name (str): Experiment name. Defaults to "default".  If it is '' then no per-experiment subdirectory is used.
```
Comment (Borda) [inline pytorch_lightning/loggers/tensorboard.py]: what is the '' ?

File: pytorch_lightning/loggers/tensorboard.py [hunk 2]
```diff
@@ -47,6 +48,26 @@ def __init__(self, save_dir, name="default", version=None, **kwargs):
...
+            return os.path.join(self.save_dir, self.name)
+
+    @property
+    def log_dir(self):
+        """ The directory for this run's tensorboard checkpoint.  By default, it is named 'version_${self.version}'
+            but it can be overridden by passing a string value for the constructor's version parameter
+            instead of None or an int"""
+        # create a pseudo standard path ala test-tube
+        log_dir = os.path.join(self.root_dir,
+                               self.version if isinstance(self.version, str) else 'version_%s' % self.version)
```
Comment (Borda) [inline pytorch_lightning/loggers/tensorboard.py]: rather use extra var eg 
```
version = self.version if isinstance(self.version, str) else f"version_{self.version}"
log_dir = os.path.join(self.root_dir, version)
```

File: pytorch_lightning/loggers/tensorboard.py [hunk 3]
```diff
@@ -29,9 +29,10 @@ class TensorBoardLogger(LightningLoggerBase):
 
     Args:
         save_dir (str): Save directory
-        name (str): Experiment name. Defaults to "default".
-        version (int): Experiment version. If version is not specified the logger inspects the save
+        name (str): Experiment name. Defaults to "default".  If it is '' then no per-experiment subdirectory is used.
+        version (int/str): Experiment version. If version is not specified the logger inspects the save
```
[the remark this answers is not in the document]
Comment (bobkemp) [reply pytorch_lightning/loggers/tensorboard.py]: Done

File: pytorch_lightning/loggers/tensorboard.py (hunk 2 above)
Comment (bobkemp) [reply pytorch_lightning/loggers/tensorboard.py]: Done

**Comment by bobkemp** [comment]
FYI, I slightly simplified my test.  It still tests as much as the existing ones but doesn't check for the tensorboard output directory being created.  Perhaps tensorboard doesn't create it immediately when saving in earlier versions.  I couldn't get it to fail locally and it works for non-minimal versions.

File: pytorch_lightning/loggers/tensorboard.py [hunk 4]
```diff
@@ -29,9 +29,11 @@ class TensorBoardLogger(LightningLoggerBase):
 
     Args:
         save_dir (str): Save directory
-        name (str): Experiment name. Defaults to "default".
-        version (int): Experiment version. If version is not specified the logger inspects the save
+        name (str): Experiment name. Defaults to "default".  If it is the empty string then no per-experiment
+        subdirectory is used.
+        version (int|str): Experiment version. If version is not specified the logger inspects the save
         directory for existing versions, then automatically assigns the next available version.
```
Comment (Borda) [inline pytorch_lightning/loggers/tensorboard.py]: this should have extra tab spacing as it belongs as a block to an argument

File: pytorch_lightning/loggers/tensorboard.py [hunk 5]
```diff
@@ -47,6 +49,26 @@ def __init__(self, save_dir, name="default", version=None, **kwargs):
         self.tags = {}
         self.kwargs = kwargs
 
+    @property
+    def root_dir(self):
+        """ Parent directory for all tensorboard checkpoint subdirectories.
+            If the experiment name parameter is None or the empty string, no experiment subdirectory is used
```
Comment (Borda) [inline pytorch_lightning/loggers/tensorboard.py]: there is no for extra tab spacing

File: pytorch_lightning/loggers/tensorboard.py (hunk 1 above)
Comment (bobkemp) [reply pytorch_lightning/loggers/tensorboard.py]: empty string

File: pytorch_lightning/loggers/tensorboard.py (hunk 4 above)
Comment (bobkemp) [reply pytorch_lightning/loggers/tensorboard.py]: Done

File: pytorch_lightning/loggers/tensorboard.py (hunk 5 above)
Comment (bobkemp) [reply pytorch_lightning/loggers/tensorboard.py]: I think it now follows the format used elsewhere.  We're getting there.

Comment (Borda) [reply pytorch_lightning/loggers/tensorboard.py]: the point/issue is that still do not have unified formatting everywhere yet...

**Pull request merged**